File: haskell/strain/src/Strain.hs
module Strain (keep, discard) where

-- Reimplement `filter`

discard :: (a -> Bool) -> [a] -> [a]
discard p xs = keep (\x -> not (p x)) xs

keep :: (a -> Bool) -> [a] -> [a]
keep p (x:xs)
  | p x = x:(keep p xs)
  | otherwise = keep p xs
keep _ [] = []

File: go/strain/strain.go
// Package strain does straining.
package strain

type Ints []int
type Lists [][]int
type Strings []string

func (i Ints) Keep(f func(int) bool) Ints {
	var ret Ints
	for _, v := range i {
		if f(v) {
			ret = append(ret, v)
		}
	}
	return ret
}

func (i Ints) Discard(f func(int) bool) Ints {
	return i.Keep(func(v int) bool { return ! f(v) })
}

func (l Lists) Keep(f func([]int) bool) Lists {
	var ret Lists
	for _, v := range l {
		if f(v) {
			ret = append(ret, v)
		}
	}
	return ret
}

func (s Strings) Keep(f func(string) bool) Strings {
	var ret Strings
	for _, v := range s {
		if f(v) {
			ret = append(ret, v)
		}
	}
	return ret
}
